sharktank/sharktank/layers/paged_attention.py

```diff
@@ -315,116 +331,201 @@ def write_timestep(
+        devices = self.pipeline_to_device_map[pipeline]
+        transformer_block_count = self.pipeline_to_block_count[pipeline]
+        block_offset = self.pipeline_to_block_offset[pipeline]
+        transformer_block_index = transformer_block_index - block_offset
 
-            index = page_id
-            index = index * transformer_block_count + transformer_block
-            index = index * self.cache_partition_count + partitions
-            index = index * self.block_seq_stride + page_offset
+        page_tables = self.unflatten_page_tables(state)
+        page_table = page_tables[pipeline]
+        page_table = page_table.flatten(0, 2)
+        assert len(cache_partitions) == self.cache_partition_count
 
-            values = ops.to(cache_partition, dtype=page_table.dtype)
-            if page_table.dtype == torch.float8_e4m3fnuz:
-                # Workaround for Torch not supporting torch.Tensor.index_copy_ for f8.
-                page_table_as_int8 = page_table.view(dtype=torch.int8)
-                values_int8 = values.view(dtype=torch.int8)
-                page_table_as_int8.index_put_(indices=(index,), values=values_int8)
-            else:
-                page_table.index_put_(indices=(index,), values=values)
+        for p, partition in enumerate(cache_partitions):
+            if not isinstance(page_table, ShardedTensor):
+                self.write_shard(
+                    page_table=page_table,
+                    cache_partition=partition,
+                    cache_partition_id=p,
+                    transformer_block_index=transformer_block_index,
+                    transformer_block_count=transformer_block_count,
+                    page_ids=page_ids,
+                )
+                continue
+
+            if (
+                isinstance(page_table, ShardedTensor)
+                and isinstance(partition, ShardedTensor)
+                and isinstance(page_ids, ReplicatedTensor)
+            ):
+
+                page_table_split = isinstance(page_table, SplitPrimitiveTensor)
+                partition_split = isinstance(partition, SplitPrimitiveTensor)
+                assert page_table_split == partition_split
+
+                page_table_shards = page_table.shards
+                partition_shards = self.deshard_like(page_table, partition)
+                page_ids_shards = self.deshard_like(page_table, page_ids)
+
+                for i in range(page_table.shard_count):
+                    device = page_table.devices[i]
+
+                    self.write_shard(
+                        page_table=page_table_shards[i],
+                        cache_partition=partition_shards[i],
+                        cache_partition_id=p,
+                        transformer_block_index=transformer_block_index,
+                        transformer_block_count=transformer_block_count,
+                        page_ids=page_ids_shards[i],
+                    )
+                continue
```

Comment:
Needs details on how this is sharded, still hard to understand. I'm also not sure why these sharding details are leaking into the PagedAttention layer, PagedAttention should simply work on a single shard and these details should be outside.